man/man1/heffalump.1
.\" Process this file with
.\" groff -man -Tascii bam-rmdup.1
.\"
.TH HEFFALUMP 1 "JULY 2017" Applications "User Manuals"
.SH NAME
heffalump \- compact storage of genotypes
.SH SYNOPSIS

.B heffalump --help

.B heffalump --version

.B heffalump hetfa
\fB\-o\fR \fIoutput.hef\fR
\fB\-r\fR \fIref.2bit\fR
\fB\-s\fR \fIsample.hetfa\fR

.BR heffalump " [" maf "|" emf "]"
\fB\-o\fR \fIoutput.hef\fR
\fB\-r\fR \fIref.2bit\fR
\fB\-R\fR \fIref-name\fR
\fB\-S\fR \fIsample-name\fR
\fIinput-file\fR ...

.BR heffalump " [" bcfin "|" vcfin ]
\fB\-o\fR \fIoutput.hef\fR
\fB\-r\fR \fIref.2bit\fR
[\fB\-D\fR|\fB-S\fR] [\fB-L\fR|\fB-H\fR]
\fIinput-file\fR...

.B heffalump bamin
\fB\-o\fR \fIoutput.hef\fR
\fB\-r\fR \fIref.2bit\fR
[\fB\-m\fR \fImin-base-qual\fR]
[\fB\-q\fR \fImin-map-qual\fR]
[\fB--snip\fR=\fIcut-length-left\fR]
[\fB--snap\fR=\fIcut-length-right\fR]
[\fB--deaminate\fR]
[\fB--stranded\fR]
\fIinput.bam\fR...

.B heffalump patch
[\fB\-w\fR \fIwidth\fR]
[\fB\-o\fR \fIoutput.hef\fR]
[\fB\-r\fR \fIref.2bit\fR]
\fB\-s\fR \fIsample.hef\fR

.B heffalump treemix
[\fB\-o\fR \fIoutput.hef\fR]
[\fB\-r\fR \fIref.2bit\fR]
[\fB\-i\fR \fIindividuals.ind\fR]
[\fB\-n\fR \fInumoutgroups\fR]
[\fB\-t\fR] [\fB-a\fR|\fB-x\fR|\fB-y\fR]
[\fB\-c\fR \fIchroms\fR]
[\fB\-R\fR \fIregions.bed\fR]

.B heffalump eigenstrat 
\fB\-o\fR \fIoutput\fR [\fB-r\fR \fIref.2bit\fR]
[\fB\-n\fR \fInumoutgroups\fR] [\fB-t\fR] [\fB-b\fR] 
[\fB\-R\fR \fIregions.bed\fR] \fIinput.hef\fR...

.B heffalump pseudo_ms 
[\fB\-r\fR \fIref.2bit\fR]
[\fB\-m\fR \fImin-informative\fR]
[\fB\-l\fR \fIblocklength\fR]
[\fB\-n\fR \fInumoutgroups\fR|\fB\-D\fR] [\fB-t\fR] [\fB-b\fR]
[\fB\-R\fR \fIregions.bed\fR] \fIinputsss.hef\fR...

.B heffalump vcfexport 
[\fB\-r\fR \fIref.2bit\fR]
[\fB\-n\fR \fInumoutgroups\fR|\fB\-D\fR] [\fB-t\fR] [\fB-b\fR] 
[\fB\-R\fR \fIregions.bed\fR] \fIinput.hef\fR...

.B heffalump kayvergence
[\fB\-r\fR \fIref.2bit\fR]
[\fB\-n\fR \fInumgood\fR]
[\fB\-J\fR \fIblocksize\fR]
[\fB\-x\fR]
[\fB\-R\fR \fIregions.bed\fR]
\fIinput.hef\fR...

.B heffalump
[\fBdstatistics\fR|\fBf4statistics\fR]
[\fB\-r\fR \fIref.2bit\fR]
[\fB\-n\fR \fInumoutgroups\fR]
[\fB\-k\fR \fInumrefpanel\fR]
[\fB\-J\fR \fIblocksize\fR]
[\fB\-x\fR] [\fB--ignore-cpg\fR]
[\fB\-R\fR \fIregions.bed\fR]
\fIinput.hef\fR...

.SH DESCRIPTION
.B heffalump
is both a file format and the program to work with it.  A heffalump file
stores the genetic differences of an individual compactly, without
losing the distinction between
.IR uncalled " and " invariant " sites."

The
.B heffalump
file format can serve as the primary storage format for a collection of
genomes;  multiple files can be merged into the input for population
genetic analysis programs relatively quickly.

The functions
.BR hetfa ", " maf ", " emf ", " bcfin ", " vcfin " and " bamin
serve to import individuals in the formats \fIhetfa\fR (e.g. Simons
Foundation diversity panel), \fImaf\fR or \fIemf\fR (e.g. whole genome
alignments of apes), high coverage \fIvcf\fR or \fIbcf\fR
(genotype-called individuals, treated as diploid), and low coverage
\fIvcf\fR or \fIbcf\fR (sampled raw reads from individuals, treated as
haploid), respectively.

The functions 
.BR treemix ", " eigenstrat ", " pseudo_ms " and " vcfexport
merge multiple
.I hef
files into one dataset useful as input for 
.BR Eigensoft ", " Treemix ", " ABLE
and a multitude of other tools, respectively.

The functions
.BR kayvergence ", " f4statistics " and " dstatistics
compute three simple kinds of relevant statistics on the fly.

.SH OPTIONS

\fBheffalump\fR is called with a command, and different commands accept
different options, see below.  Commands are case-insensitive and a
prefix is accepted as long as it is unambiguous.  A few options are
global or recurring:

.TP
.B --version
Prints the version number and exits.

.TP
.B --help
Prints the list of available commands;  the list of commands is also
printed if an unknown, ambigous or no command at all is given.  The
valid subcommands also accept
.B --help
and print appropriate usage information.  Some commands come without
documentation;  these should be treated a dangerous and/or bad ideas.

.TP
\fB\-r\fR, \fB--reference\fR \fIfile.2bit\fR
Specifies a \fI2bit\fR file containing the reference genome.  A
reference genome is always needed when importing anything into \fIhef\fR
format, and the location of the reference is remembered.  Only \fIhef\fR
files created with respect to the same reference can be processed
together.  When operating on \fIhef\fR files, the same reference is
usually needed, and then \fB\-r\fR needs to be specified explicitly only
if the reference isn't found automatically.  A few operations (plain
\fItreemix\fR output, \fIdstatistics\fR, \fIf4statistics\fR) can work
without access to the reference.

\fBheffalump\fR stores genomes as a list of difference to a reference.
When importing things like \fIhetfa\fR, the reference is clearly needed to
find differences.  Other formats like \fImaf\fR have the reference built
in, but \fBheffalump\fR still needs a reference genome to establish the
correct order of reference scaffolds.  For \fIbam\fR, even the order is
implied, but here \fBheffalump\fR needs the reference to remove unwanted
scaffolds and to provide rudimentary error detection.

.TP
\fB\-n\fR, \fB--numoutgroups\fR \fInum\fR
Specifies the number of outgroups.  This option is common to operations
that merge multiple \fIhef\fR files and produce a set of variants.  It
states that the first \fInum\fR inputs are to be treated as outgroups,
and the default is one outgroup.  A site where only outgroups, but no
other samples, differ from the reference, does not constitute a variant
and is omitted from the output.  (Specifying the correct number of
outgroups is not merely of cosmetic effect, it tends to save a
substantial amount of time, too.)


.SH COMMANDS

.SS Hetfa

\fBheffalump hetfa\fR
\fB\-o\fR \fIoutput.hef\fR
\fB\-r\fR \fIref.2bit\fR
\fB\-s\fR \fIsample.hetfa\fR

Imports one \fIhetfa\fR file into \fIhef\fR format.  Note that the
filter file that usually accompanies the \fIhetfa\fR file is not
supported, because I consider this kind of arbitrary filter harmful.

.TP
\fB\-o\fR, \fB--output\fR=\fIfile.hef\fR
Write output to \fIfile.hef\fR.

.TP
\fB\-r\fR, \fB--reference\fR=\fIref.2bit\fR
Read the reference genome from \fIref.2bit\fR.

.TP
\fB\-s\fR, \fB--sample\fR=\fIsample.hfa\fR
Read the sample from \fIsample.hfa\fR.  The input must be in hetfa
format (nucleotide fasta, with IUPAC ambiguity codes for heterozygous
calls, Ns or dashes for uncalled sites, optionally Qs for sites that
match te reference) and can optionally be compressed with \fBgzip\fR or
\fBbgzip\fR.

.SS Maf/Emf

\fBheffalump\fR [\fBmaf\fR|\fBemf\fR]
\fB\-o\fR \fIoutput.hef\fR
\fB\-r\fR \fIref.2bit\fR
\fB\-R\fR \fIreference-species\fR
[\fB\-S\fR \fIsample-species\fR|\fB\-A\fR \fIsample-species\fR]
\fIinput-file\fR...

Imports one genome from a set of \fImaf\fR or \fIemf\fR files into
\fIhef\fR format.  The input files can appear in any order and the
fragments extracted from them are reordered and sanity checked.
Duplicated or inconsistent fragments elicit warnings, but
\fBheffalump\fR still imports them by resolving ambiguities arbitrarily.

.TP
\fB\-o\fR, \fB--output\fR=\fIfile.hef\fR
Write output to \fIfile.hef\fR.

.TP
\fB\-r\fR, \fB--reference\fR=\fIref.2bit\fR
Read the reference genome from \fIref.2bit\fR.

.TP
\fB\-R\fR, \fB--ref-species\fR=\fIname\fR
Sets the name of the reference species to \fIname\fR; this should be the
same species whose genome is supplied as reference.  For \fImaf\fR, the
\fIname\fR must be spelled exactly as it appears in the input files.  For
\fIemf\fR, it must be the scientific name, written like "homo,sapiens".

.TP
\fB\-S\fR, \fB--sample-species\fR=\fIname\fR
Sets the name of the species to be imported to \fIname\fR, see above for
the correct spelling.

.TP
\fB\-A\fR, \fB--ancestor-of\fR=\fIname\fR
Imports the ancestor of the reference species and species \fIname\fR,
see above for the correct spelling.  Only \fIemf\fR files contain
(inferred) ancestral information, so this option is invalid when
importing \fImaf\fR.

.SS Bcfin/Vcfin

\fBheffalump\fR [\fBbcfin\fR|\fBvcfin\fR]
\fB\-o\fR \fIoutput.hef\fR
\fB\-r\fR \fIref.2bit\fR
[\fB\-D\fR|\fB\-S\fR|\fB--good-patches\fR|\fB--poor-patches\fR]
[\fB\-L\fR|\fB\-H\fR] 
\fIinput-file\fR...

Imports the first individual from a set of \fIbcf\fR or bgzip-compressed
\fIvcf\fR files.  The input files are concatenated, so they must be
sorted and appear in the correct order.  The only difference between
\fIbcf\fR and \fIvcf\fR is that \fIbcf\fR is markedly more robust and
efficient.

.TP
\fB\-o\fR, \fB--output\fR=\fIfile.hef\fR
Write output to \fIfile.hef\fR.

.TP
\fB\-r\fR, \fB--reference\fR=\fIref.2bit\fR
Read the reference genome from \fIref.2bit\fR.

.TP
\fB\-D\fR, \fB--dense\fR
The input is "dense", meaning it is expected to contain a record for
every site.  Sites absent from the input are assumed to be uncalled,
that is, missing data.
.B heffalump
will normally detect if the input is dense and act accordingly.

.TP
\fB\-S\fR, \fB--sparse\fR
The input is "sparse", meaning it is expected to contain only records
for variant sites.  Sites absent from the input are assumed to match the
reference.
.B heffalump
will normally detect if the input is sparse and act accordingly.

.TP
\fB--good-patches\fR=\fIpatches.bed\fR
The input is "sparse", but of such low quality that the missing
positions cannot always be assumed to match the reference.  The file
.I patches.bed
lists high quality regions.
.B heffalump
will assume the missing positions within high quality regions match the
reference, while missing positions outside high quality regions are not
called.  Put another way,
.B heffalump
behaves as if
.B --dense
was specified for the regions in the bed file, but
.B --sparse
for the remaining parts of the genome.  This option is only intended to
repair badly prepared VCF input; if you need
.BR --patchwork ,
it would be better to fix the problem upstream.

.TP
\fB--poor-patches\fR=\fIpatches.bed\fR
The input is "sparse", but of such low quality that the missing
positions cannot always be assumed to match the reference.  The file
.I patches.bed
lists low quality regions.
.B heffalump
will assume the missing positions outside low quality regions match the
reference, while missing positions within low quality regions are not
called.  Put another way,
.B heffalump
behaves as if
.B --sparse
was specified for the regions in the bed file, but
.B --dense
for the remaining parts of the genome.  This option is only intended to
repair badly prepared VCF input; if you need
.BR --poor-patches ,
it would be better to fix the problem upstream.

.TP
\fB\-H\fR, \fB--high\fR
The input is of "high coverage", meaning it is expected that genotype
calling is reliable.  Such input is imported unmodified.  This is the
default mode.

.TP
\fB\-L\fR, \fB--low\fR
The input is of "low coverage", meaning it is not expected that both
alleles of the samples could be observed.  The homozygous diploid calls
in such input are transformed into haploid calls to reflect the low
information content.

.SS Bamin

\fBheffalump bamin\fR
\fB\-o\fR \fIoutput.hef\fR
\fB\-r\fR \fIref.2bit\fR
[\fB\-q\fR, \fB--min-qual\fR=\fIqual\fR]
[\fB\-m\fR, \fB--min-mapq\fR=\fImapq\fR]
[\fB--snip\fR \fIsnip\fR]
[\fB--snap\fR \fIsnap\fR]
[\fB--stranded\fR]
\fIinput.bam\fR...

Crudely imports a set of \fIbam\fR files into \fIhef\fR format by
randomly selecting one of the crossing reads at each site and using its
raw base as the haploid call.  Input files must be sorted and are
merged.

A multitude of sampling schemes is available and was used in preliminary
experiments, but none is fit for actual use.  Please refer to the source
code to figure out how to enable them and what exactly they do.

.TP
\fB\-o\fR, \fB--output\fR=\fIfile.hef\fR
Write output to \fIfile.hef\fR.

.TP
\fB\-r\fR, \fB--reference\fR=\fIref.2bit\fR
Read the reference genome from \fIref.2bit\fR.  Only scaffolds present
in the reference are imported and arranged into the correct order,
missing scaffolds are stored as uncalled.  It is reported as an error if
no scaffold could be imported.

.TP
\fB\-q\fR, \fB--min-qual\fR=\fIqual\fR
Ignore bases with a quality score below \fIqual\fR.

.TP
\fB\-m\fR, \fB--min-mapq\fR=\fImapq\fR
Ignore reads with a map quality below \fImapq\fR.

.TP
\fB--stranded\fR
Apply "strand sensitive sampling":  sample only As and Gs from reads
aligning in the forward direction, sample only Ts and Cs from reads
aligning in the reverse direction.  This is expected to remove the
effect of deamination without introducing any bias.

.SS Patch

\fBheffalump patch\fR
[\fB\-o\fR \fIoutput.hfa\fR]
[\fB\-r\fR \fIref.2bit\fR]
\fB\-s\fR \fIsample.hef\fR

Applies the variants in one \fIhef\fR file to the reference, thus
regenerating the original \fIhetfa\fR file.

.TP
\fB\-o\fR, \fB--output\fR=\fIfile.hfa\fR
Write output to \fIfile.hfa\fR.  By default, output is written to
\fIstdout\fR.

.TP
\fB\-r\fR, \fB--reference\fR=\fIref.2bit\fR
Read the reference genome from \fIref.2bit\fR.

.TP
\fB\-s\fR, \fB--sample\fR=\fIsample.hef\fR
Read the sample from \fIsample.hef\fR.

.TP
\fB\-w\fR, \fB--width\fR=\fInum\fR
Break output lines after \fInum\fR columns.  The default is 50.

.SS pseudo_ms

\fBheffalump\fR [\fBpseudo_ms\fR]
[\fB\-r\fR \fIref.2bit\fR]
[\fB\-m\fR \fImin-informative\fR]
[\fB\-l\fR \fIblockLength\fR]
[\fB\-n\fR \fInumoutgroups\fR|\fB\-D\fR]
[\fB\-R\fR \fIregions.bed\fR]
[\fB\-t\fR] [\fB\-b\fR]
\fIinput.hef\fR...

Merges multiple \fIhef\fR files, splits the resulting variants into
biallelic variants, and prints the result in \fIpseudo_ms\fR format.

.TP
\fB\-r\fR, \fB--reference\fR=\fIref.2bit\fR
Read the reference genome from \fIref.2bit\fR.

.TP
\fB\-n\fR, \fB--numoutgroups\fR=\fInum\fR
The first \fInum\fR inputs are outgroups.  The output is "sparse" in the
sense that it will contain only sites with variants.

.TP
\fB\-m\fR, \fB--min-informative\fR=\fInum\fR
The minimum \fIpercentage\fR of informative-sites in a block to consider it as a
valid block.

.TP
\fB\-l\fR, \fB--blockLength\fR=\fInum\fR
Length of each block.

.TP
\fB\-D\fR, \fB--dense\fR
Produce a "dense" output file which will contain at least one record for
each site, even if no individual has a variant.  The only sites
suppressed from the output are those where no individual is called.

.TP
\fB\-t\fR, \fB--only-transversions\fR
Output only transversion variants.

.TP
\fB\-b\fR, \fB--only-biallelic\fR
Output only biallelic variants.  Multi-allelic variants are always split
into multiple biallelic variants, this options only decides whether to
output them.

.TP
\fB\-R\fR, \fB--regions\fR=\fIregions.bed\fR
Restrict output to variants in the regions found in \fIregions.bed\fR,
which should be a \fIbed\fR file.

.SS Treemix

\fBheffalump treemix\fR
[\fB\-o\fR \fIoutput.tmx.gz\fR]
[\fB\-r\fR \fIref.2bit\fR]
[\fB\-i\fR \fIindividuals.ind\fR]
[\fB\-n\fR \fInumoutgroups\fR]
[\fB\-R\fR \fIregions.bed\fR]
[\fB\-t\fR] [\fB\-a\fR|\fB\-x\fR|\fB\-y\fR]
[\fB\-c\fR \fIchroms\fR]
\fIinput.hef\fR...

Merges multiple \fIhef\fR files, splits the resulting variants into
biallelic variants, collates observations by population, and prints the
result in a format digestible by \fBtreemix\fR.  

.TP
\fB\-o\fR, \fB--output\fR=\fIfile.tmx.gz\fR
Write output to \fIfile.tmx.gz\fR in gzip compressed treemix format.  By
default, output is written to \fIstdout\fR.

.TP
\fB\-r\fR, \fB--reference\fR=\fIref.2bit\fR
Read the reference genome from \fIref.2bit\fR.  The reference is needed
to decode the legacy \fIheffalump\fR format and to implement the
\fB\-R\fR, \fB\-a\fR, \fB\-x\fR, \fB\-y\fR, and \fB\-c\fR options.
Usually, the reference is found automatically without explicitly
specifying it.

.TP 
\fB\-i\fR, \fB--individuals\fR=\fIfile.ind\fR
Read individuals from \fIfile.ind\fR,  which lists individuals and the
populations they belong to in the same format as used by
\fBEigenstrat\fR.  \fBheffalump\fR will search the list of input files
for those whose base names match the individual names, and it will
produce one column in the output for each distinct population.  If no
individual file is specified, each individual becomes a singleton
population.

.TP
\fB\-n\fR, \fB--numoutgroups\fR=\fInum\fR
The first \fInum\fR inputs (on the command line, not in the individuals
file) are outgroups.

.TP
\fB\-t\fR, \fB--transversions\fR
Output only transversion variants.

.TP
\fB\-a\fR, \fB--autosomes\fR
Process only the autosomes, i.e. everything that matches 
.IR ^(chr)?[0-9]+[a-z]?$ .

.TP
\fB\-x\fR, \fB--x-chromosome\fR
Process only the X chromosome, i.e. everything that matches 
.IR ^(chr)?X$ .

.TP
\fB\-y\fR, \fB--y-chromosome\fR
Process only the Y chromosome, i.e. everything that matches 
.IR ^(chr)?Y$ .

.TP
\fB\-c\fR, \fB--chromosomes\fR=\fIregex\fR
Process only target sequences that match \fIregex\fR.

.TP
\fB\-R\fR, \fB--regions\fR=\fIregions.bed\fR
Restrict output to variants in the regions found in \fIregions.bed\fR,
which should be a \fIbed\fR file.  If both the \fB\-R\fR option and one
of the \fB\-a\fR, \fB\-x\fR, \fB\-y\fR, or \fB\-c\fR options are given,
only variants in the bed file's regions on the selected chromosomes are
processed.

.SS Eigenstrat/Vcfexport

\fBheffalump\fR [\fBeigenstrat\fR|\fBvcfexport\fR]
[\fB\-o\fR \fIoutput\fR]
[\fB\-r\fR \fIref.2bit\fR]
[\fB\-n\fR \fInumoutgroups\fR|\fB\-D\fR]
[\fB\-R\fR \fIregions.bed\fR]
[\fB\-t\fR] [\fB\-b\fR]
\fIinput.hef\fR...

Merges multiple \fIhef\fR files, splits the resulting variants into
biallelic variants, and prints the result in \fIeigenstrat\fR or
\fIvcf\fR format, respectively.

.TP
\fB\-o\fR, \fB--output\fR=\fIbasename\fR
For \fBeigenstrat\fR, write output to \fIbasename.geno\fR and
\fIbasename.snp\fR.  Streaming or compressing this format is not
possible due to the confused design of the \fBEigenstrat\fR package.
For \fBvcfexport\fR, output is always written to stdout.

.TP
\fB\-r\fR, \fB--reference\fR=\fIref.2bit\fR
Read the reference genome from \fIref.2bit\fR.

.TP
\fB\-n\fR, \fB--numoutgroups\fR=\fInum\fR
The first \fInum\fR inputs are outgroups.  The output is "sparse" in the
sense that it will contain only sites with variants.

.TP
\fB\-D\fR, \fB--dense\fR
Produce a "dense" output file which will contain at least one record for
each site, even if no individual has a variant.  The only sites
suppressed from the output are those where no individual is called.

.TP
\fB\-t\fR, \fB--only-transversions\fR
Output only transversion variants.

.TP
\fB\-b\fR, \fB--only-biallelic\fR
Output only biallelic variants.  Multi-allelic variants are always split
into multiple biallelic variants, this options only decides whether to
output them.

.TP
\fB\-R\fR, \fB--regions\fR=\fIregions.bed\fR
Restrict output to variants in the regions found in \fIregions.bed\fR,
which should be a \fIbed\fR file.

.SS Kayvergence/Dstatistics/F4statistics

\fBheffalump\fR [\fBkayvergence\fR|\fBdstatistics\fR|\fBf4statistics\fR]
[\fB\-r\fR \fIref.2bit\fR]
[\fB\-n\fR \fInumoutgroups\fR]
[\fB\-k\fR \fInumrefpanel\fR]
[\fB\-R\fR \fIregions.bed\fR]
[\fB\-J\fR \fIblocksize\fR]
[\fB\-t\fR] [\fB\--ignore-cpg\fR]
\fIinput.hef\fR...

Merges multiple \fIhef\fR files and computes either "Kayvergence ratios"
or "D-statistics" and "F4-statistics" on suitable combinations of the
inputs.  Output is formatted as a human readable table and sent to
stdout.

"Kayvergence ratio" is a crude estimate of population history computed
from shared variants.  \fBheffalump\fR computes the "kayvergence ratio"
for any combination of two "good genomes" and one "bad genome" from the
input; the reference itself counts as an additional "good genome".  The
p-value applies to the hypothesis that the "bad genome" is an outgroup
to the two "good genomes".

"D-statistics" tests the hypothesis that the allele sharing of four
populations is consistent with a clean tree shaped history.
\fBheffalump\fR computes the "D-statistics" for any combination of one
"outgroup", two individuals from the "reference panel" and one "sample"
from the input, where the assumed population history separates the
sample and the outgroup from the reference panel.

"F4-statistics" is a proportion of shared changes between two sets of

src/heffalump.hs
import Bio.Prelude
import Control.Monad.Trans.Resource
import Paths_heffalump                  ( version )
import Streaming
import System.Console.GetOpt
import System.IO

import qualified Data.ByteString.Builder         as B
import qualified Data.ByteString.Char8           as B
import qualified Data.ByteString.Lazy.Char8      as L
import qualified Data.ByteString.Streaming       as S
import qualified Data.IntMap                     as I
import qualified Streaming.Prelude               as Q

import Bamin
import Eigenstrat
import Emf
import Genome
import Lump
import SillyStats
import Stretch ( main_dumppatch )
import Treemix
import Util
import Vcf
import VcfOut
import ShameMS

main :: IO ()
main = do
    as0 <- getArgs
    case as0 of
        [] -> usage
        a:as -> case [ m | (k,(m,_)) <- mains
                     , map toLower a `isPrefixOf` map toLower k ] of
            [m] -> m as
            _   -> usage
  where
    usage = do
        pn <- getProgName
        let kl = maximum $ map (length . fst) mains
        hPutStrLn stderr . unlines $ "Usage:" :
              [ "  "++pn++' ':k++replicate (kl-length k) ' '++"  -- "++h
              | (k,(_,h)) <- mains ]

    mains = let z a b c = (a,(b,c)) in
        [ z "hetfa"        main_hetfa             "Import hetfa file"
        , z "maf"          main_maf               "Import one genome from maf files"
        , z "emf"          main_emf               "Import one genome from emf (Compara)"
        , z "bcfin"       (main_xcf conf_bcf)     "Import one genome from bcf files"
        , z "vcfin"       (main_xcf conf_vcf)     "Import one genome from vcf files"
        , z "bamin"        main_bam               "Import crudely from bam files"
        , z "patch"        main_patch             "Make a hetfa file by patching the reference"
        , z "treemix"      main_treemix           "Merge heffalumps into Treemix format"
        , z "eigenstrat"   main_eigenstrat        "Merge heffalumps into Eigenstrat format"
        , z "vcfexport"    main_vcfout            "Merge heffalumps into vcf format"
        , z "pseudo_ms"    mainShameMSout         "Merge heffalumps into pseudo_ms format"
        , z "twobitinfo"   main_2bitinfo          "List reference sequences"
        , z "twobittofa"   main_2bittofa          "Extract Fasta from 2bit"
        , z "fatotwobit"   main_fato2bit          "Convert Fasta to 2bit"
        , z "kayvergence"  main_kayvergence       "Compute Kayvergence ratios"
        , z "f4statistics"(main_patterson WantF4) "Compute Patterson's F4 and D"
        , z "dstatistics" (main_patterson NoF4)   "Compute Patterson's D"
        , z "yaddayadda"   main_yaddayadda        "Compute Yadda-Yadda-counts"
        , z "dumppatch"    main_dumppatch         "(debugging aid)"
        , z "dumplump"     main_dumplump          "(debugging aid)"
        , z "--help"      (const usage)           "list commands and exit"
        , z "--version"    main_version           "print version and exit" ]

main_version :: [String] -> IO ()
main_version _ = hPutStrLn stderr $ showVersion version

data ConfImportGen = ConfImportGen {
    conf_imp_reference  :: FilePath,
    conf_imp_output     :: FilePath,
    conf_imp_sample     :: FilePath }
  deriving Show

defaultImportConf :: ConfImportGen
defaultImportConf = ConfImportGen (error "no reference specified")
                                  (error "no output file specified")
                                  (error "no sample file specified")


opts_hetfa :: [ OptDescr (ConfImportGen -> IO ConfImportGen) ]
opts_hetfa =
    [ Option "o" ["output"] (ReqArg set_output "FILE") "Write output to FILE (.hef)"
    , Option "r" ["reference"] (ReqArg set_ref "FILE") "Read reference from FILE (.2bit)"
    , Option "s" ["sample"] (ReqArg set_sample "FILE") "Read sample from FILE (.hetfa)" ]
  where
    set_output a c = return $ c { conf_imp_output    = a }
    set_ref    a c = return $ c { conf_imp_reference = a }
    set_sample a c = return $ c { conf_imp_sample    = a }


main_hetfa :: [String] -> IO ()
main_hetfa args = do
    ConfImportGen{..} <- parseOpts defaultImportConf (mk_opts "hetfa" [] opts_hetfa) args
    refs <- readTwoBit conf_imp_reference

    runResourceT $ S.writeFile conf_imp_output $
            importHetfa refs $ readSampleFa conf_imp_sample


-- | Import Hetfa by diffing it against a reference.  We must read the
-- Hetfa in the order of the file (Fasta doesn't have an index), but
-- write it in the order of the reference.  So we parse a chromosome at
-- a time, sort the diffs in memory and dump it at the end.
--
-- Chromosomes in input that don't have a match in the reference are
-- silently ignored.  Chromsomes that don't show up in the input are
-- emitted as empty sequences.  Only if no sequence matches do we
-- produce an error message.
importHetfa :: MonadIO m => RefSeqs -> Stream (FastaSeq m) m r -> S.ByteString m r
importHetfa ref smps = S.mwrap $ do
    (map1,r) <- fold_1 I.empty smps

    when (I.null map1) $ error
            "Found only unexpected sequences.  Is this the right reference?"

    return $ do S.fromLazy $ B.toLazyByteString $ encodeHeader ref
                forM_ [0 .. length (rss_chroms ref) - 1] $
                        \i -> S.fromLazy $ unpackLump $ I.findWithDefault noLump i map1
                pure r
  where
    enc2 :: MonadIO m => Int -> S.ByteString m r -> m (Of PackedLump r)
    enc2 i sq = encodeLumpToMem $ diff2 (rss_seqs ref !! i) sq <* Q.yield Break

    fold_1 :: MonadIO m => I.IntMap PackedLump -> Stream (FastaSeq m) m r -> m (I.IntMap PackedLump, r)
    fold_1 !acc s = inspect s >>= \case
        Left r -> return (acc,r)
        Right (FastaSeq nm sq) -> case elemIndex nm (rss_chroms ref) of
            Nothing -> S.effects sq >>= fold_1 acc
            Just  i -> do lump :> s' <- enc2 i sq
                          liftIO $ hPrint stderr (i, L.length $ unpackLump lump)
                          fold_1 (I.insert i lump acc) s'

data ConfMaf = ConfMaf
    { conf_maf_output :: FilePath
    , conf_maf_reference :: FilePath
    , conf_maf_ref_species :: B.ByteString
    , conf_maf_oth_species :: B.ByteString }

opts_maf :: [ OptDescr ( ConfMaf -> IO ConfMaf ) ]
opts_maf =
    [ Option "o" ["output"]     (ReqArg set_output "FILE") "Write output to FILE (.hef)"
    , Option "r" ["reference"]     (ReqArg set_ref "FILE") "Read reference from FILE (.2bit)"
    , Option "R" ["ref-species"]  (ReqArg set_from "NAME") "Set reference species to NAME"
    , Option "S" ["sample-species"] (ReqArg set_to "NAME") "Set sample species to NAME" ]
  where
    set_output a c = return $ c { conf_maf_output = a }
    set_ref    a c = return $ c { conf_maf_reference = a }
    set_from   a c = return $ c { conf_maf_ref_species = fromString a }
    set_to     a c = return $ c { conf_maf_oth_species = fromString a }

conf_maf :: ConfMaf
conf_maf = ConfMaf (error "no output file specified")
                   (error "no reference specified")
                   (error "no reference species specified")
                   (error "no sample species specified")

main_maf :: [String] -> IO ()
main_maf args = do
    (maffs,ConfMaf{..}) <- parseFileOpts conf_maf "maf" "[maf-file...]" opts_maf args
    ref <- readTwoBit conf_maf_reference
    withFile conf_maf_output WriteMode $ \ohdl ->
        L.hPut ohdl . encodeGenome =<<
           foldM (\g f -> withInputFile f $
                            parseMaf (conf_maf_ref_species, conf_maf_oth_species)
                                     (rss_chroms ref) g . decomp . S.fromHandle)
                 emptyGenome (if null maffs then ["-"] else maffs)


data ConfPatch = ConfPatch {
    conf_patch_width      :: Int64,
    conf_patch_output     :: (Handle -> IO ()) -> IO (),
    conf_patch_reference  :: Either String FilePath,
    conf_patch_sample     :: FilePath }

defaultPatchConf :: ConfPatch
defaultPatchConf = ConfPatch 50 ($ stdout)
                             (Left    "no reference specified")
                             (error "no sample file specified")


opts_patch :: [ OptDescr (ConfPatch -> IO ConfPatch) ]
opts_patch =
    [ Option "o" ["output"] (ReqArg set_output "FILE") "Write output to FILE (.hetfa)"
    , Option "r" ["reference"] (ReqArg set_ref "FILE") "Read reference from FILE (.2bit)"
    , Option "s" ["sample"] (ReqArg set_sample "FILE") "Read sample from FILE (.hef)"
    , Option "w" ["width"]  (ReqArg  set_width  "NUM") "Set width of FastA to NUM (50)" ]
  where
    set_output "-" c = return $ c { conf_patch_output    = ($ stdout) }
    set_output  a  c = return $ c { conf_patch_output    = withFile a WriteMode }
    set_ref     a  c = return $ c { conf_patch_reference = Right a }
    set_sample  a  c = return $ c { conf_patch_sample    =       a }
    set_width   a  c = (\n   -> c { conf_patch_width     =       n }) <$> readIO a

main_patch :: [String] -> IO ()
main_patch args = do
    ConfPatch{..} <- parseOpts defaultPatchConf (mk_opts "patch" [] opts_patch) args
    withFile conf_patch_sample ReadMode $ \hdli -> do
        (mref, raw) <- getRefPath $ decomp $ S.fromHandle hdli
        ref <- readTwoBit $ either (`fromMaybe` mref) id conf_patch_reference

        conf_patch_output $ \hdlo ->
            patchFasta hdlo conf_patch_width
                (rss_chroms ref) (rss_seqs ref) (decode (Right ref) raw)

main_dumplump :: [String] -> IO ()
main_dumplump [ref,inf] = do rs <- readTwoBit ref
                             withFile inf ReadMode $ debugLump . decode (Right rs) . decomp . S.fromHandle
main_dumplump [  inf  ] =    withFile inf ReadMode $ debugLump . decode (Left "no reference given") . decomp . S.fromHandle
main_dumplump     _     =    hPutStrLn stderr "Usage: dumplump [foo.hef]"


patchFasta :: Handle                    -- ^ output handle
           -> Int64                     -- ^ line width
           -> [B.ByteString]            -- ^ chromosome names
           -> [() -> RefSeq]            -- ^ reference sequences
           -> Stream (Of Lump) IO r     -- ^ input heffalump
           -> IO r
patchFasta hdl wd = p1
  where
    p1 [    ]     _  p = Q.effects p
    p1      _ [    ] p = Q.effects p
    p1 (c:cs) (r:rs) p = do hPutStrLn hdl $ '>' : B.unpack c
                            p2 (p1 cs rs) 0 (patch (r ()) p)

    p2 k l | l == wd = \f -> hPutChar hdl '\n' >> p2 k 0 f
    p2 k l = inspect >=> \case
        Left p            -> when (l>0) (L.hPutStrLn hdl L.empty) >> k p
        Right (Short c f) -> hPutChar hdl c >> p2 k (succ l) f
        Right (Long  s f)
            | L.null s    -> p2 k l f
            | otherwise   -> do let (u,v) = L.splitAt (wd-l) s
                                L.hPutStr hdl u
                                p2 k (l + L.length u) (wrap $ Long v f)

